include/aws/s3/s3_client.h

```diff
@@ -198,6 +200,9 @@ struct aws_s3_meta_request_options {
 
     /* Callback for when the meta request has completely cleaned up. */
     aws_s3_meta_request_shutdown_fn *shutdown_callback;
+
+    /* If present, then overrides HTTP Request host header */
```

Comment:
That confuses me without read the implementation. We don't use the URI to update the host header right?
Also, what if the URI is set, and host is set as well? Who takes in charge? _Should it be error if they both set and different with each other?_ TBH, I don't know what's the right answer here, maybe it's fine to let the URI to take in charge as user set it specifically.
Suggestion:
```
/** 
 * Optional.
 * The endpoint for the meta request to connect to. If not set, the "Host" header from `message` will be used instead.
 **/
```

Reply:
In the Java (above JNI) I replace the host header if the uri is provided. The reason is - if the host header is incorrect (or - as in my testing I used constant string "INVALID") then the connection fails. I looked but was not able to easily see why. It made sense that the Java should force the host header to match the host in the uri - so that's the approach I took.  It's clear that some other path through the code relies on the host header, not that endpoint arg that is passed around. I felt  it would be too extensive of a change to try to find and replace that functionality.

Reply:
Well, if the host header is "incorrect/invalid", the user who pass the http request to java should be the one to fix it, right? I don't think overwrite the host header is a right way to do even for java side.

Reply:
I would add one observation: replacing the host header because of the URI makes that change visible to the caller of the `makeMetaRequest` method of `S3Client.java`. This might be surprising to the caller - they pass host header 'foo' and it changes to 'bar' on return. If that's undesirable, other options are 1) clone/modify the `HttpRequest` in the `makeMetaRequest` method and pass the clone to the JNI so the caller's original is untouched 2) error if uri hostname != http request host header 3) disallow host headers in the http request if URI is also provided, 4) other? Fundamentally, allowing the caller to supply URI creates the potential for a clash with http request host header so maybe some additional handling is indicated...

Reply:
I agree on those besides

>uri hostname non null && HttpRequest -> host header null = ok - C code constructs host header from uri hostname for lower layers in the call chain

I don't like the client to modify the passed in request. Our S3 client is not responsible for correcting the request user passed in. So, I'd just error out at that case.

Reply:
Pushed updates in this repo and for https://github.com/awslabs/aws-crt-java/pull/435

Reply:
two trivial things:
- Can you rename this to `endpoint`? I think it's more informative than `uri`.
- Can we update the documentation to something like:
```
/** 
 * Optional. 
 * The endpoint for the meta request to connect to. 
 * If not set, the tls settings from client will be used to determine the port and host header from `message` determine the host to connect.
 * Note: need to meet the host header from `message`
 **/
```

Reply:
changed in latest commit